File: src/main/java/com/learning/java/classobj/TemperatureAlarm.java
package com.learning.java.classobj;

/* Это Java-класс который будет выглядеть одной строкой в Scala-классе */


public class TemperatureAlarm {

	private final double temp;

	public TemperatureAlarm(double temp) {
		this.temp = temp;
	}

	public double getTemp() {
		return temp;
	}

	@Override
	public String toString() {
		return "TemperatureAlarm [temp=" + temp + "]";
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		long temp;
		temp = Double.doubleToLongBits(this.temp);
		result = prime * result + (int) (temp ^ (temp >>> 32));
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		TemperatureAlarm other = (TemperatureAlarm) obj;
		if (Double.doubleToLongBits(temp) != Double.doubleToLongBits(other.temp))
			return false;
		return true;
	}
}

File: src/main/scala/com/learning/scala/classobj/TemperatureAlarm.scala
package com.learning.scala.classobj

/* Scala: как использовать 'case-класс' ? */

/*
 * Ключевое слово 'case' - заставляет Scala-компилятор создавать такие дополнительные методы как: 'hashCode()', 'equals()', 'toString()'.
 * (Он также создает дополнительно много строительных методов. Большую часть времени, эти другие методы не используются непосредственно разработчиком).
 * Они существуют чтобы обеспечить RunTime-поддержку. Scala предлагает следующее:
 * - для того чтобы 'case-класс' были использованы в поиске выражений...
 */
case class TemperatureAlarm(temp: Double)

object TemperatureAlarmTest {

  def main(args: Array[String]) {

    // 'Case-классы' могут быть созданы без необходимости повторного ключевого слова 'case'
    val alarm = TemperatureAlarm(99.9)
    println(alarm.toString()) /* (метод предоставляемый компилятором) */
    println(alarm.hashCode()) /* (метод предоставляемый компилятором) */

    /* Scala-конструктор поиска по шаблону работает только в классе 'case': */
    val msgAlarm = alarm match {
      case TemperatureAlarm(temp) => "Tripped at temp " + temp
      case _ => "No match"
    }
    println( msgAlarm )
  }
}
